File: apis/csharp/routine-api/util/DataWriter.cs
using System;
using System.IO;
using routine_api.api;
using Thrift.Protocol;
using Thrift.Transport;

namespace routine_api.util
{
    /// <summary>
    /// Writes data to a Stream in the Thrift Binary Protocol.
    /// </summary>
    public class DataWriter : IDisposable, IPipeWriter
    {
        private bool _disposed;

        private readonly TStreamTransport _outPipe;
        private readonly TProtocol _outProto;

        /// <inheritdoc />
        /// <summary>
        /// Builds a DataWriter that writes data to a File. A Filestream will be opened to access a named pipe.
        /// </summary>
        /// <param name="outPipePath">The path to the file to open the stream</param>
        public DataWriter(string outPipePath) : this(File.OpenWrite(outPipePath))
        {
        }

        /// <summary>
        /// Builds a DataWriter that writes data to a stream. This uses the provided stream directly.
        /// </summary>
        /// <param name="outPipeStream">The stream to use</param>
        public DataWriter(Stream outPipeStream)
        {
            _outPipe = new TStreamTransport(null, outPipeStream);
            _outProto = new TBinaryProtocol(_outPipe);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (_disposed) return;
            if (disposing)
            {
                _outPipe?.Dispose();
            }

            _disposed = true;
        }

        /// <summary>
        /// Writes an instance of type T to the stream.
        /// </summary>
        /// <param name="instance">The instance to write</param>
        /// <exception cref="IOException">If there is an error reading</exception>  
        /// <typeparam name="T">The type of the instance. Needs to be a Thrift generated type</typeparam>
        public void Store<T>(T instance) where T : TBase
        {
            instance.Write(_outProto);
            _outPipe.Flush();
        }

        /// <summary>
        /// Writes a byte array to the stream.
        /// </summary>
        /// <param name="data">The data to write</param>
        /// <exception cref="TTransportException">If there is an error reading</exception>  
        public void Store(byte[] data)
        {
            _outPipe.Write(data, 0, data.Length);
        }

        /// <summary>
        /// Write a string to the stream.
        /// </summary>
        /// <param name="str">The string to write</param>
        /// <exception cref="TTransportException">If there is an error reading</exception>  
        public void Store(string str)
        {
            _outProto.WriteString(str);
        }

        /// <summary>
        /// Writes an integer to the stream.
        /// </summary>
        /// <param name="i">The integer to write</param>
        /// <exception cref="TTransportException">If there is an error reading</exception>  
        public void Store(int i)
        {
            _outProto.WriteI32(i);
        }
    }
}

File: routine-api/src/main/java/at/enfilo/def/routine/util/DataWriter.java
package at.enfilo.def.routine.util;

import at.enfilo.def.routine.api.IPipeWriter;
import org.apache.thrift.TBase;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TIOStreamTransport;
import org.apache.thrift.transport.TTransportException;

import java.io.*;

public class DataWriter implements Closeable, IPipeWriter {
	private final TIOStreamTransport outPipe;
	private final TProtocol outProto;

	public DataWriter(File outPipe) throws IOException {
		this(new FileOutputStream(outPipe));
	}

	public DataWriter(OutputStream os) {
		this.outPipe = new TIOStreamTransport(os);
		this.outProto = new TBinaryProtocol(this.outPipe);
	}

	@Override
	public void close() {
		outPipe.close();
	}

	/**
	 * Store data to pipe.
	 * @param t - data to store
	 * @param <T> - type of data
	 * @throws IOException
	 */
	public <T extends TBase> void store(T t) throws TException {
		t.write(outProto);
		outPipe.flush();
	}


	/**
	 * Write/store bytes to pipe.
	 *
	 * @param data - bytes to store
	 * @throws IOException
	 * @throws TTransportException
	 */
	public void store(byte[] data) throws TTransportException {
		outPipe.write(data, 0, data.length);
		outPipe.flush();
	}


	/**
	 * Write/store an objective string to pipe.
	 *
	 * @param str - string to store
	 * @throws TException
	 */
	public void store(String str) throws TException {
		outProto.writeString(str);
		outPipe.flush();
	}


	/**
	 * Write store an objective integer to pipe.
	 *
	 * @param i
	 * @throws TException
	 */
	public void store(int i) throws TException {
		outProto.writeI32(i);
		outPipe.flush();
	}
}
